packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+    if (bottomSheetHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
+
+    return new Offset(fabX, fabY);
+  }
+}
+
+class _EndFloatFabPositioner extends FloatingActionButtonPositioner {
+  const _EndFloatFabPositioner();
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    // Compute the x-axis offset.
+    double fabX;
+    assert(scaffoldGeometry.textDirection != null);
+    switch (scaffoldGeometry.textDirection) {
+      case TextDirection.rtl:
+        fabX = _kFloatingActionButtonMargin + scaffoldGeometry.horizontalFloatingActionButtonPadding;
+        break;
+      case TextDirection.ltr:
+        fabX = scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.floatingActionButtonSize.width - _kFloatingActionButtonMargin - scaffoldGeometry.horizontalFloatingActionButtonPadding;
+      break;
+    }
+
+    // Compute the y-axis offset.
+    final double contentBottom = scaffoldGeometry.contentBottom;
+    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
+    final double fabHeight = scaffoldGeometry.floatingActionButtonSize.height;
+    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
+
+    double fabY = contentBottom - fabHeight - _kFloatingActionButtonMargin;
+    if (snackBarHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - snackBarHeight - fabHeight - _kFloatingActionButtonMargin);
+    if (bottomSheetHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
+
+    return new Offset(fabX, fabY);
+  }
+}
+
+/// A snapshot of a transition between two [FloatingActionButtonPositioner]s.
+///
+/// [ScaffoldState] uses this to seamlessly change transition animations
+/// when a running [FloatingActionButtonPositioner] transition is interrupted by a new transition.
+@immutable
+class _TransitionSnapshotFabPositioner extends FloatingActionButtonPositioner {
+  
+  const _TransitionSnapshotFabPositioner(this.begin, this.end, this.animator, this.progress);
+
+  final FloatingActionButtonPositioner begin;
+  final FloatingActionButtonPositioner end;
+  final FloatingActionButtonAnimator animator;
+  final double progress;
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    return animator.getOffset(
+      begin: begin.getOffset(scaffoldGeometry), 
+      end: end.getOffset(scaffoldGeometry), 
+      progress: progress,
+    );
+  }
+
+  @override
+  String toString() {
+    return '$runtimeType(begin: $begin, end: $end, progress: $progress)';
+  }
+}
+
+/// Geometry information for [Scaffold] components after layout is finished.
 ///
-/// To get a [ValueNotifier] for the scaffold geometry of a given
-/// [BuildContext], use [Scaffold.geometryOf].
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+/// 
+/// For information about the [Scaffold]'s geometry that is used as part of
```

Comment:
The two layout geometry classes need better explanations.  Why do both exist? When can the geometry information be used?